sshd-core/src/main/java/org/apache/sshd/common/BaseBuilder.java

```diff
@@ -70,6 +70,7 @@
      */
     public static final List<BuiltinCiphers> DEFAULT_CIPHERS_PREFERENCE = Collections.unmodifiableList(
             Arrays.asList(
+                    // BuiltinCiphers.cc20p1305_openssh, // TODO: enable by default when BouncyCastle available
```

Comment:
Why this? The AES256 ciphers also may not be available, but are unconditionally in this list.

Reply:
I'll probably just copy in an implementation of ChaCha here for a fallback when BC isn't available.

Reply:
I may be missing something here... but this BuiltinCiphers code already can have ciphers in there that have isSupported() == false. So I wonder why we'd need to have a ChaCha implementation in sshd itself. Either some provider provides it, or not.

I would expect the ChaCha cipher never to be proposed or used if it isn't supported. Having it supported only if BC is available would be fine by me. (We seem to require BC >= 1.68. If there's any such BC version that doesn't provide ChaCha, then the feature test would need to be more precise.)

Don't know what the stance of the other maintainers is on having more crypto code inside sshd. AFAIK the only crypto code in sshd is the OpenBSD bcrypt implementation needed to handle OpenSSH encrypted keys. That one was copied because it was felt that having another dependency on an external artifact just for one class was to be avoided, and the inclusion was cleared beforehand here at Apache _and_ with the Eclipse legal team. (The latter only because I wanted it to be used in Eclipse; including it here if then Eclipse legal had objected would have only complicated matters.)

Reply:
The copy of ChaCha I'm thinking of is one I wrote in the first place based on the public domain reference implementation. However, if there's a clean way to simply require BouncyCastle to enable this cipher, then we can avoid any additional cipher code being copied. As I mentioned before, the main problem here is that OpenSSH decided to use a pre-standardized version of ChaCha20-Poly1305 which works slightly differently than the one bundled with Java 11+.

Reply:
If I understand you right, BC provides this pre-standard ChaCha cipher. But SecurityUtils gives no clean way to ask for a "ChaCha" cipher from a particular provider ("BC"). So that's what I was missing. BuiltinCiphers.cc20p1305_openssh does check whether BC is registered, but in ChaChaCipher when calling SecurityUtils.getCipher() there's no guarantee you get the one from BC. Bypass SecurityUtils and ask via Cipher.getInstance("ChaCha", "BC") directly? Or indeed include your own implementation.

Reply:
Yeah, that was the problem. I'm adding in an implementation for this to simplify things.